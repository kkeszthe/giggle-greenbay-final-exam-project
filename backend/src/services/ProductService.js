import { userService } from '../dependencies/dependencyInjection';

export class ProductService {
  constructor(productRepo, errorCodes) {
    this.product = productRepo;
    this.user = userService;
    this.errorCodes = errorCodes;
  }

  validateURL(string) {
    var pattern = new RegExp(
      '^(https?:\\/\\/)?' + // protocol
        '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|' + // domain name
        '((\\d{1,3}\\.){3}\\d{1,3}))' + // OR ip (v4) address
        '(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*' + // port and path
        '(\\?[;&a-z\\d%_.~+=-]*)?' + // query string
        '(\\#[-a-z\\d_]*)?$',
      'i'
    ); // fragment locator
    return !!pattern.test(string);
  }

  async getSellableProducts() {
    const products = await this.product.getSellableProducts();
    return { products };
  }
  async getById({ productId }) {
    return (await this.product.getById({ productId }))[0];
  }

  async add({ product_name, description, photo_url, price, seller_id }) {
    if (!product_name || !description || !seller_id)
      throw new Error(this.errorCodes.missingParam);

    if (!price || !(Number.isInteger(price) && price > 0))
      throw new Error(this.errorCodes.invalidPrice);
    if (!photo_url || this.validateURL(photo_url))
      throw new Error(this.errorCodes.invalidURL);
    const productId = (
      await this.product.add({
        product_name,
        description,
        photo_url,
        price,
        seller_id,
      })
    ).insertId;
    return await this.getById({ productId });
  }

  async sellProduct({ productId, userId }) {
    if (!productId || !userId) throw new Error(this.errorCodes.missingParam);
    const product = await this.getById({ productId });
    const buyer = await this.user.getById({ userId });
    if (!product) throw new Error(this.errorCodes.invalidProductId);
    if (product.price > buyer.balance)
      throw new Error(this.errorCodes.invalidBalance);
    if (product.buyer_id) throw new Error(this.errorCodes.usedUserId);
    await this.product.sellProduct({ productId, userId });
    await this.user.updateBalance({ userId, amount: product.price * -1 });
    await this.user.updateBalance({
      userId: product.seller_id,
      amount: product.price,
    });
    return await this.getById({ productId });
  }
}
