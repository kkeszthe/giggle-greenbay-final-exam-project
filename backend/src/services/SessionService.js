import jwt from 'jsonwebtoken';

export class SessionService {
  constructor(userService, errorCodes) {
    this.user = userService;
    this.errorCodes = errorCodes;
  }

  validateParams({ username, password }) {
    if (!username && !password)
      throw new Error(this.errorCodes.missingUsernameAndPassword);
    if (!username) throw new Error(this.errorCodes.missingUsername);
    if (!password) throw new Error(this.errorCodes.missingPassword);
  }

  getToken({ userId }) {
    return jwt.sign({ userId }, process.env.JWT_SECRET || 'secret');
  }

  async login({ username, password }) {
    this.validateParams({ username, password });
    const user = await this.user.getByUsernameAndPassword({
      username,
      password,
    });
    if (user.length === 0)
      throw new Error(this.errorCodes.invalidUsernameAndPassword);
    return {
      token: this.getToken(user.userId),
      userId: user.id,
      username: user.username,
      photo_url: user.photo_url,
      balance: user.balance,
    };
  }

  verifyToken({ token }) {
    if (!token) throw new Error(this.errorCodes.missingToken);
    try {
      const { userId } = jwt.verify(token, process.env.JWT_SECRET || 'secret');
      return { userId };
    } catch (error) {
      throw new Error(this.errorCodes.invalidToken);
    }
  }
}
