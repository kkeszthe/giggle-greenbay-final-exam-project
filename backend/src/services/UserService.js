export class UserService {
  constructor(userRepo, errorCodes) {
    this.user = userRepo;
    this.errorCodes = errorCodes;
  }

  validateParams({ username, password }) {
    if (!username && !password)
      throw new Error(this.errorCodes.missingUsernameAndPassword);
    if (!username) throw new Error(this.errorCodes.missingUsername);
    if (!password) throw new Error(this.errorCodes.missingPassword);
    if (password.length < 8) throw new Error(this.errorCodes.invalidPassword);
  }

  validateURL(string) {
    var pattern = new RegExp(
      '^(http://www.|https://www.|http://|https://)?[a-z0-9]+([-.]{1}[a-z0-9]+)*.[a-z]{2,5}(:[0-9]{1,5})?(/.*)?'
    );
    return pattern.test(string);
  }

  async add({ username, password, photo_url }) {
    this.validateParams({ username, password });
    if (photo_url || this.validateURL(photo_url))
      throw new Error(this.errorCodes.invalidURL);
    const defaultBalance = 0;
    const userId = (
      await this.user.add({
        username,
        password,
        photo_url,
        balance: defaultBalance,
      })
    ).insertId;

    return {
      id: userId,
      username: username,
    };
  }

  async getByUsernameAndPassword({ username, password }) {
    if (!username) throw new Error(this.errorCodes.missingUsername);
    if (!password) throw new Error(this.errorCodes.missingPassword);
    return (await this.user.getAuthentication({ username, password }))[0];
  }

  async getById({ userId }) {
    return (await this.user.getById({ userId }))[0];
  }

  async updateBalance({ userId, amount }) {
    if (!amount) throw new Error(this.errorCodes.missingParam);
    const balance = ((await this.getById({ userId })).balance += parseInt(
      amount
    ));
    await this.user.updateBalance({ userId, balance });
    return await this.getById({ userId });
  }
}
