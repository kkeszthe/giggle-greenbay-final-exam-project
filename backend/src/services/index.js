export * from './UserService';
export * from './SessionService';
export * from './ProductService';
