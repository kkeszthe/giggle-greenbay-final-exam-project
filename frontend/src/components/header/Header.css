header.MuiAppBar-colorPrimary {
  background-color: seagreen;
}

span.balance {
  margin: 10px;
  font-weight: bolder;
}
