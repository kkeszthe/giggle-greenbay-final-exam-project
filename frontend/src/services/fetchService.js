import { env } from '../env';

export const generalFetch = async (endpoint, { method, token, body }) => {
  const result = await fetch(`${env.BACKEND_URL}/api/${endpoint}`, {
    method,
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      token,
    },
    body: JSON.stringify(body),
  });
  return result.json();
};
