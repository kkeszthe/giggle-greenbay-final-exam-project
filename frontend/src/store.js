import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';

import { userReducer } from './reducers/userReducer';
import { productsReducer } from './reducers/productsReducer';
import { tokenReducer } from './reducers/tokenReducer';
import { errorReducer } from './reducers/errorReducer';

function root(state = {}, action) {
  return {
    user: userReducer(state.user, action),
    products: productsReducer(state.troops, action),
    token: tokenReducer(state.token, action),
    error: errorReducer(state.error, action),
  };
}

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const store = createStore(
  root,
  /* preloadedState, */ composeEnhancers(applyMiddleware(thunk))
);

export { store };
