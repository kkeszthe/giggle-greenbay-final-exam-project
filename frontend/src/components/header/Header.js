import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { useHistory } from 'react-router-dom';

import { makeStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import Avatar from '@material-ui/core/Avatar';
import AccountCircle from '@material-ui/icons/AccountCircle';
import MenuItem from '@material-ui/core/MenuItem';
import Menu from '@material-ui/core/Menu';

import './Header.css';
import TopUpBox from '../user/TopUpBox';
import SellProductBox from '../products/SellProductBox';
import Error from '../Error';
import { getUserAction } from '../../actions/UserActions';

const useStyles = makeStyles(theme => ({
  root: {
    flexGrow: 1,
  },
  menuButton: {
    marginRight: theme.spacing(2),
  },
  title: {
    flexGrow: 1,
  },
}));

const Header = ({ user, token, get }) => {
  const history = useHistory();
  const classes = useStyles();
  const [anchorEl, setAnchorEl] = React.useState(null);
  const open = Boolean(anchorEl);

  const handleMenu = event => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  const [openSellProduct, setOpenSellProduct] = React.useState(false);
  const [openTopUp, setOpenTopUp] = React.useState(false);

  const handleClickOpenSellProduct = () => {
    setOpenSellProduct(true);
  };
  const handleClickOpenTopUp = () => {
    setOpenTopUp(true);
  };
  useEffect(() => {
    if (token) {
      get();
    }
  }, [token, get]);

  return (
    <div className={classes.root}>
      <AppBar position="static">
        <Toolbar>
          <Typography variant="h3" className={classes.title}>
            GreenBay
          </Typography>

          {
            <div>
              <IconButton
                aria-label="account of current user"
                aria-controls="menu-appbar"
                aria-haspopup="true"
                onClick={handleMenu}
                color="inherit"
              >
                {token ? (
                  <Avatar alt="Profile Picture" src={user.photo_url} />
                ) : (
                  <AccountCircle />
                )}
              </IconButton>
              {token ? (
                <>
                  <Typography variant="overline">{user.username}</Typography>
                  <Typography variant="overline" className={'balance'}>
                    {user.balance + ' GBD'}
                  </Typography>
                </>
              ) : null}

              <Menu
                id="menu-appbar"
                anchorEl={anchorEl}
                anchorOrigin={{
                  vertical: 'top',
                  horizontal: 'right',
                }}
                keepMounted
                transformOrigin={{
                  vertical: 'top',
                  horizontal: 'right',
                }}
                open={open}
                onClose={handleClose}
              >
                {token ? (
                  [
                    <MenuItem
                      onClick={() => {
                        handleClose();
                        handleClickOpenTopUp();
                      }}
                      key={1}
                    >
                      Top-up account
                    </MenuItem>,
                    <MenuItem
                      onClick={() => {
                        handleClose();
                        handleClickOpenSellProduct();
                      }}
                      key={2}
                    >
                      Sell product
                    </MenuItem>,
                  ]
                ) : (
                  <MenuItem
                    onClick={() => {
                      handleClose();
                      history.push('/login');
                    }}
                  >
                    Login
                  </MenuItem>
                )}
              </Menu>
            </div>
          }
        </Toolbar>
      </AppBar>
      <TopUpBox open={openTopUp} setOpen={setOpenTopUp} />
      <SellProductBox open={openSellProduct} setOpen={setOpenSellProduct} />
      <Error />
    </div>
  );
};

Header.propTypes = {
  user: PropTypes.object,
};

const mapStateToProps = ({ user, token }) => {
  return { user, token };
};

const mapDispatchToProps = dispatch => {
  return {
    get: () => {
      return dispatch(getUserAction());
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Header);
