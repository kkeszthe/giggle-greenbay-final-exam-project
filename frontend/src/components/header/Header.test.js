import React from 'react';
import { Provider } from 'react-redux';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import configureStore from 'redux-mock-store';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
jest.mock('../.././services/fetchService');
import { generalFetch } from '../../services/fetchService';

import Header from './Header';
import { store } from '../../store';

const history = createMemoryHistory({
  initialEntries: ['/'],
});
const mockStore = configureStore([]);

generalFetch.mockImplementation(() =>
  Promise.resolve({
    id: 11,
    username: 'sal8',
    photo_url:
      'https://www.nordmet.se/wp-content/uploads/2017/11/Siluett-med-spelarprofilbild-2.jpg',
    balance: 100,
  })
);

let container;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it('renders Header without crashing without token', async () => {
  const mockedStore = mockStore({
    user: {},
    products: [],
    product: {},
    token: '',
  });
  mockedStore.dispatch = jest.fn();

  await act(async () => {
    render(
      <Provider store={mockedStore}>
        <Router history={history}>
          <Header />
        </Router>
      </Provider>,
      container
    );
  });
});

it('renders Header without crashing with token', async () => {
  Storage.prototype.getItem = jest.fn(key => {
    return 'dummy';
  });

  await act(async () => {
    render(
      <Provider store={store}>
        <Router history={history}>
          <Header />
        </Router>
      </Provider>,
      container
    );
  });
});

it('matches snapshot without token', async () => {
  const mockedStore = mockStore({
    user: {},
    products: [],
    product: {},
    token: '',
  });
  mockedStore.dispatch = jest.fn();

  await act(async () => {
    render(
      <Provider store={mockedStore}>
        <Router history={history}>
          <Header />
        </Router>
      </Provider>,
      container
    );
  });
  expect(container.innerHTML).toMatchSnapshot();
});

it('matches snapshot with token and user', async () => {
  Storage.prototype.getItem = jest.fn(key => {
    return 'dummy';
  });

  await act(async () => {
    render(
      <Provider store={store}>
        <Router history={history}>
          <Header />
        </Router>
      </Provider>,
      container
    );
  });
  expect(container.innerHTML).toMatchSnapshot();
});
