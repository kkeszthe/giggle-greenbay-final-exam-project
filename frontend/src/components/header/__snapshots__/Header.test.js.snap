// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot with token and user 1`] = `"<div class=\\"makeStyles-root-10\\"><header class=\\"MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4\\"><div class=\\"MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters\\"><h3 class=\\"MuiTypography-root makeStyles-title-12 MuiTypography-h3\\">GreenBay</h3><div><button class=\\"MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"account of current user\\" aria-controls=\\"menu-appbar\\" aria-haspopup=\\"true\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button></div></div></header><div></div></div>"`;

exports[`matches snapshot without token 1`] = `"<div class=\\"makeStyles-root-7\\"><header class=\\"MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4\\"><div class=\\"MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters\\"><h3 class=\\"MuiTypography-root makeStyles-title-9 MuiTypography-h3\\">GreenBay</h3><div><button class=\\"MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"account of current user\\" aria-controls=\\"menu-appbar\\" aria-haspopup=\\"true\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button></div></div></header><div></div></div>"`;
